// k-lunch/index.js
const cheerio = require('cheerio')
const request = require('request')
const codefinder = require('./excel/codefinder')

const pad = d => (d < 10) ? '0' + d.toString() : d.toString()

module.exports.getLunch = (form, callback, options) => {
  const ymdArr = [form.year, form.month, form.day]
  const ymd = ymdArr.map(v => pad(v)).join('')

  if(options) {
    const schoolName = form.name

    if(options.autoCode) {
      const schoolCode = codefinder.getCode(schoolName)
      form.code = schoolCode
    }

    if(options.autoDomain) {
      const domain = codefinder.getDomain(schoolName)
      form.domain = domain
    }
  }

  const reqForm = {
    "schYmd": ymd,
    "schMmealScCode": form.time,
    "insttNm": form.name,
    "schulCode": form.code,
    "schulKndScCode": pad(form.phase),
    "schulCrseScCode": form.phase
  }

  const URL = `http://stu.${form.domain}.go.kr/sts_sci_md01_001.do`

  request({
    headers: {'content-type' : 'application/x-www-form-urlencoded'},
    url: URL,
    form: reqForm
  }, (err, res_post, body) => {
    if(err) {
      return callback(err, null)
    } else {
      const $ = cheerio.load(body, {decodeEntities: false})
      const mixedDate = ymdArr.map(v => pad(v)).join('.')

      let index = 0
      let data = ''
      let dataArr = []
      let nutsArr = []
      let output = []

      $('table > thead > tr').each(function() {
        let rows = $(this).find('th')
        
        for(let i = 1; i < 8; i++) {
          let rowText = rows.eq(i).text();
          rowText.search(mixedDate) !== -1 ? index = i : i
        }
      });

      const findByRow = function(n) {
        let obj = {}

        $('table > tbody').each(function() {
          let lunchTr = $(this).find('tr').eq(n)
          obj.key = lunchTr.find('th').html()
          obj.value = lunchTr.find('td').eq(index - 1).html()
          return false
        })

        return obj
      }

      data = findByRow(1).value

      if(data === null || $('table > thead > tr > td').text() === '자료가 없습니다.') {
        return callback("There's no data in table.", null)
      } else {
        dataArr = data.split('<br>')
        dataArr.pop()

        dataArr.forEach(e => {
          let intIndex = e.search(/\d/)
          let dataSchema = {
            menu: '',
            allergyInfo: ''
          }

          if(intIndex === -1) {
            dataSchema.menu = e
          } else {
            dataSchema.menu = e.substring(0, intIndex).replace(/-/g, "")
            dataSchema.allergyInfo = e.substring(intIndex, e.length - 1).split('.')
          }
          
          output.push(dataSchema)
        });

        if(!output[0]) {
          return callback("There's no data in table.", null)
        } else {
          return callback(null, output)
        }
      }
    }
  });
}

module.exports.getNutrients = (form, callback, options) => {

  const ymdArr = [form.year, form.month, form.day]
  const ymd = ymdArr.map(v => pad(v)).join('')

  if(options) {
    const schoolName = form.name

    if(options.autoCode === true) {
      const schoolCode = codefinder.getCode(schoolName)
      form.code = schoolCode
    }

    if(options.autoDomain === true) {
      const domain = codefinder.getDomain(schoolName)
      form.domain = domain
    }
  }

  const reqForm = {
    "schYmd": ymd,
    "schMmealScCode": form.time,
    "insttNm": form.name,
    "schulCode": form.code,
    "schulKndScCode": pad(form.phase),
    "schulCrseScCode": form.phase
  }

  const URL = `http://stu.${form.domain}.go.kr/sts_sci_md01_001.do`

  request({
    headers: {'content-type' : 'application/x-www-form-urlencoded'},
    url: URL,
    form: reqForm
  }, (err, res_post, body) => {
    if(err) {
      return callback(err, null)
    } else {
      const $ = cheerio.load(body, {decodeEntities: false})
      const mixedDate = ymdArr.map(v => pad(v)).join('.')

      let index = 0
      let output = {}

      $('table > thead > tr').each(function() {
        let rows = $(this).find('th')
        
        for(let i = 1; i < 8; i++) {
          let rowText = rows.eq(i).text();
          if(rowText.search(mixedDate) !== -1){
          	index = i
        	}
        }
      });

      const findByRow = function(n) {
        let obj = {}

        $('table > tbody').each(function() {
          let lunchTr = $(this).find('tr').eq(n)
          obj.key = lunchTr.find('th').html()
          obj.value = lunchTr.find('td').eq(index - 1).html()
          return false
        })

        return obj
      }

      const getNuts = () => {
        let obj = {}
        for(let i = 44; i < 54; i++) {
          let rowObj = findByRow(i)
          if(rowObj.value !== '') {
            obj[rowObj.key] = rowObj.value
          }
        }
        return obj
      }

      const isObjectEmpty = obj => Object.keys(obj).length === 0 && obj.constructor === Object

      output = getNuts()

      if(isObjectEmpty(output)) {
        return callback("There's no data in table.", null)
      } else {
        return callback(null, output)
      }   
    }
  })
}
